product_packaging_container_deposit/models/product_product.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2023 Camptocamp (<https://www.camptocamp.com>).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+from odoo.tools import groupby
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def get_product_container_deposit_quantities(self, qty, forced_packaging=False):
+        """
+        Returns the quantity of deposit per packaging level for a given product quantity
+        {
+            "PL1": (CP1, QTY),
+            "PLn": (CPn, QTYn)
+        }
+        """
+        self.ensure_one()
+        pack_qties = {}
+        if qty > 0:
+            for plevel, packs in groupby(
+                self.packaging_ids, lambda p: p.packaging_level_id
+            ):
+                # Sort by forced_packaging, fitting packagings, biggest packaging
+                packs.sort(key=lambda p: (p != forced_packaging, p.qty < qty, -p.qty))
```

Comment:
You must sort before the call to groupby because groupby works only on sorted data

Reply:
this is not the groupby from itertools :wink: